examples/object-oriented-cpp/PineAPPL.hpp

```diff
@@ -0,0 +1,242 @@
+    pineappl_grid *raw;
+
+    /**
+     * @brief constructor
+     * @param lumi luminosity
+     * @param orders orders
+     * @param bin_limits bin limits
+     * @param key_val additional informations
+     */
+    Grid(const Lumi& lumi, const std::vector<Order>& orders, const std::vector<double>& bin_limits, const KeyVal& key_val) {
+        // cast orders
+        const std::size_t n_orders = orders.size();
+        std::vector<std::uint32_t> raw_orders;
+        for (const Order& order : orders) {
+            raw_orders.push_back(order.alphas);
+            raw_orders.push_back(order.alpha);
+            raw_orders.push_back(order.logxir);
+            raw_orders.push_back(order.logxif);
+        }
+        this->raw = pineappl_grid_new(lumi.raw, n_orders, raw_orders.data(), bin_limits.size() - 1, bin_limits.data(), key_val.raw);
+    }
+
+    /** @brief destructor */
+    ~Grid(){ pineappl_grid_delete(this->raw); }
+
+    /**
+     * @brief Number of orders
+     * @return number of orders
+     */
+    std::size_t order_count() const {
+        return pineappl_grid_order_count(this->raw);
+    }
+
+    /**
+     * @brief Number of bins
+     * @return number of bins
+     */
+    std::size_t bin_count() const {
+        return pineappl_grid_bin_count(this->raw);
+    }
+
+    /**
+     * @brief Fill grid for the given parameters
+     * @param x1 first momentum fraction
+     * @param x2 second momentum fraction
+     * @param q2 scale
+     * @param order order index
+     * @param observable observable value
+     * @param lumi luminosity index
+     * @param weight weight
+     */
+    void fill(const double x1, const double x2, const double q2, const std::size_t order, const double observable, const std::size_t lumi, const double weight) const {
+        pineappl_grid_fill(this->raw, x1, x2, q2, order, observable, lumi, weight);
+    }
+
+    /**
+     * @brief perform a convolution of the grid with PDFs
+     * @param pdg_id hadron ID
+     * @param pdf PDF
+     * @param xi_ren renormalization scale variation
+     * @param xi_fac factorization scale variation
+     * @param order_mask order mask
+     * @param lumi_mask luminosity mask
+     * @return prediction for each bin
+     */
+    std::vector<double> convolute_with_one(const std::int32_t pdg_id, LHAPDF::PDF* pdf, const double xi_ren = 1.0, const double xi_fac = 1.0, const std::vector<bool>& order_mask = {}, const std::vector<bool>& lumi_mask = {}) const {
+        // prepare LHAPDF stuff
+        auto xfx = [](std::int32_t id, double x, double q2, void* pdf) {
+            return static_cast <LHAPDF::PDF*> (pdf)->xfxQ2(id, x, q2);
+        };
+        auto alphas = [](double q2, void* pdf) {
+            return static_cast <LHAPDF::PDF*> (pdf)->alphasQ2(q2);
+        };
+        // cast order_mask
+        std::unique_ptr<bool[]> raw_order_mask;
+        if (order_mask.size() > 0) {
+            raw_order_mask = std::unique_ptr<bool[]> (new bool[this->order_count()]);
```

Comment:
Here you can use `order_mask.size()` instead of `this->order_count()`. If they're not the same you will have a problem anyway.